shared/domains/chelonia/chelonia.js

```diff
@@ -270,9 +270,10 @@ export default (sbp('sbp/selectors/register', {
   },
   // 'chelonia/contract' - selectors related to injecting remote data and monitoring contracts
   // TODO: add an optional parameter to "retain" the contract (see #828)
-  'chelonia/contract/sync': function (contractIDs: string | string[]): Promise<*> {
+  'chelonia/contract/sync': async function (contractIDs: string | string[], syncDetector?: string): Promise<*> {
     const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
-    return Promise.all(listOfIds.map(contractID => {
+    syncDetector && sbp('okTurtles.data/set', syncDetector, true)
+    await Promise.all(listOfIds.map(contractID => {
```

Comment:
This is one way to do it, but I think there's another, simpler way, that doesn't require passing in `syncDetector` at all.

After line 274 we have `listOfIds` available, which tell us which contracts are about to be sync'd, and inside of the `group.js` and `chatroom.js` contracts, we are passed the `contractID` (or other contractIDs via other variables, like `data.chatRoomID`).

We can rewrite the contracts to directly check if their contractID is being sync'd, and then, I think, you will no longer need to create a string to pass in to `syncDetector`.

So, inside of `'chelonia/_init'`, you can add:

```js
this.currentSyncs = {}
```

Then, instead of modifying `'chelonia/contract/sync'`, inside of `'chelonia/private/in/syncContract'`, you can add next to `sbp('okTurtles.events/emit', CONTRACT_IS_SYNCING, contractID, true)`:

```js
this.currentSyncs[contractID] = true
```

And at the end (next to the other `CONTRACT_IS_SYNCING`), add:

```js
this.currentSyncs[contractID] = false
```

And finally, create a new selector in this file (`chelonia.js`): `'chelonia/contract/isSyncing'`:

```js
'chelonia/contract/isSyncing': function (contractID: string): ?boolean {
  return this.currentSyncs[contractID]
}
```

And then everywhere in the contracts replace the checks for strings like `'READY_TO_JOIN_CHATROOM'` with calls to `'chelonia/contract/isSyncing'` and pass in the `contractID`. `contractID` is passed in as a parameter to all `process` functions and all `sideEffect` functions.

Reply:
@taoeffect, I tried this and finally found that this could not be solution.
Because of that we use that `okTurtles.data/set` for checking if the user is joining group or chatroom contract.
Actually, **joining** is quite different from **syncing**. We can sync the contracts when we login in another device.
So it breaks many test cases. I will add this `chelonia/contract/isSyncing` but don't think it's good fit.

Reply:
Using `chelonia/contract/isSyncing`, I have removed many `okTurtles.data/set` and `okTurtles.data/get` calls.
It is really awesome...

Reply:
Awesome! 😄 🎉